virttest/qemu_devices/qcontainer.py

```diff
@@ -3105,16 +3105,23 @@ def _emulator_setup(binary, extra_options=None):
             if tpm_version in ('2.0',):
                 setup_cmd += ' --tpm2'
 
-            tpm_path = os.path.join(tmp_dir, '%s_%s_tpm_state_%s' %
-                                    (self.vmname, name, random_str))
-            if os.path.exists(tpm_path):
-                shutil.rmtree(tpm_path, ignore_errors=True)
-            os.makedirs(tpm_path)
-            setup_cmd += " --tpm-state %s" % tpm_path
+            tpm_state_path = params.get(
+                "tpm_state_path", os.path.join(
+                    tmp_dir, '%s_%s_tpm_state_%s' % (
+                        self.vmname, name, random_str)))
+            tpm_overwrite = params.get_boolean("tpm_overwrite")
+            if os.path.exists(tpm_state_path):
+                if tpm_overwrite:
```

Comment:
`if os.path.exists(tpm_state_path) and tpm_overwrite:`

Reply:
There are 4 scenarios here:

|"exist" + "tpm_overwrite"|"exist" + "not tpm_overwrite"|"not exist" + "tpm_overwrite"|"not exist" + "not tpm_overwrite"|
|:-----------------------:|:---------------------------:|:---------------------------:|:-------------------------------:|
|      shutil.rmtree      |                             |         os.makedirs         |           os.makedirs           |

Only when the path exists and requires `tpm_overwrite` then we need to remove the path, if use `os.path.exists(tpm_state_path) and tpm_overwrite`, the next check is `elif not os.path.exists(tpm_state_path)`, because `"exist" + "not tpm_overwrite"` should be skipped. Actually, if the path already exists, we can keep it here, to test `tpm_overwrite` can overwrite it, but currently seems we don't have such requirement, so the current judgment is a bit cumbersome.

Reply:
Oh, I see, there is a judgment hidden here. 
```
if os.path.exists(tpm_state_path) and tpm_overwrite:
    print('b')
elif not tpm_overwrite:
    skip
else:  
    print('a')
```
The current code is ok for me. @YongxueHong  do you have any good suggestions?

Reply:
Or maybe we add a note for later reading?

```
if os.path.exists(tpm_state_path):
    if tpm_overwrite:  # skip cases of 'os.path.exists' and 'not tpm_overwrite'
```

Reply:
> There are 4 scenarios here:
> "exist" + "tpm_overwrite" 	"exist" + "not tpm_overwrite" 	"not exist" + "tpm_overwrite" 	"not exist" + "not tpm_overwrite"
> shutil.rmtree 		os.makedirs 	os.makedirs
> 
> Only when the path exists and requires `tpm_overwrite` then we need to remove the path, if use `os.path.exists(tpm_state_path) and tpm_overwrite`, the next check is `elif not os.path.exists(tpm_state_path)`, because `"exist" + "not tpm_overwrite"` should be skipped. Actually, if the path already exists, we can keep it here, to test `tpm_overwrite` can overwrite it, but currently seems we don't have such requirement, so the current judgment is a bit cumbersome.

Hi, for the first scenario that if the condition of `"path.exist" + "tpm_overwrite"` is `True`, then we remove the `tpm_state_path`, but the line 3118 `setup_cmd += " --tpm-state %s" % tpm_state_path` looks like run fail.
CC @zhenyzha  @zhenyzha

Reply:
Yes, a good catch. But if users set `tpm_overwrite = yes`, and without setting a specific TPM state path(`tpm_state_path `), so why do they do this?
Anyway, this is an issue, and after considering again, I think only creating the state path when it does not exist is enough as I mentioned in the above comment, swtpm has the ability to overwrite an existing TPM device.